og_ui/src/Plugin/Field/FieldFormatter/GroupSubscribeFormatter.php

```diff
@@ -0,0 +1,253 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\og_ui\Plugin\Field\FieldFormatter\GroupSubscribeFormatter.
+ */
+
+namespace Drupal\og_ui\Plugin\Field\FieldFormatter;
+
+
+use Drupal\Core\Access\AccessResult;
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Field\FieldItemListInterface;
+use Drupal\Core\Field\FormatterBase;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Routing\RedirectDestinationTrait;
+use Drupal\Core\Url;
+use Drupal\og\Og;
+use Drupal\og\OgAccess;
+use Drupal\og\OgGroupAudienceHelper;
+use Drupal\user\Entity\User;
+use Drupal\user\EntityOwnerInterface;
+
+/**
+ * Plugin implementation for the OG subscribe formatter.
+ *
+ * @FieldFormatter(
+ *   id = "og_ui_group_subscribe",
+ *   label = @Translation("OG Group subscribe"),
+ *   description = @Translation("Display OG Group subscribe links."),
+ *   field_types = {
+ *     "og_group"
+ *   }
+ * )
+ */
+class GroupSubscribeFormatter extends FormatterBase {
+
+  use RedirectDestinationTrait;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function viewElements(FieldItemListInterface $items, $langcode) {
+    $elements = [];
+
+    $entity = $items->getEntity();
+    $account = User::load(\Drupal::currentUser()->id());
+    $field_name = $this->getSetting('field_name');
+
+    // Entity is not a group.
+    if (!Og::isGroup($entity->getEntityTypeId(), $entity->bundle())) {
+      return [];
+    }
+
+    if (($entity instanceof EntityOwnerInterface) && ($entity->getOwnerId() == $account->id())) {
+      // User is the group manager.
+      $elements[0] = [
+        '#type' => 'html_tag',
+        '#tag' => 'span',
+        '#attributes' => [
+          'title' => $this->t('You are the group manager'),
+          'class' => ['group', 'manager'],
+        ],
+        '#value' => $this->t('You are the group manager'),
+      ];
+
+      return $elements;
+    }
+
+    if (Og::isMember($entity, $account, [OG_STATE_ACTIVE, OG_STATE_PENDING])) {
+      if (OgAccess::userAccess($entity, 'unsubscribe', $account)) {
+        $link['title'] = $this->t('Unsubscribe from group');
+        $link['url'] = Url::fromRoute('og_ui.unsubscribe', ['entity_type_id' => $entity->getEntityTypeId(), 'entity_id' => $entity->id()]);
+        $link['class'] = ['unsubscribe'];
+      }
+    }
+    else {
+      if (Og::isMember($entity, $account, [OG_STATE_BLOCKED])) {
+        // If user is blocked, they should not be able to apply for
+        // membership.
+        return [];
+      }
+
+      // Check if user can subscribe to the field.
+      if (empty($field_name) && $audience_field_name = og_get_best_group_audience_field('user', $account, $entity_type, $bundle)) {
```

Comment:
@pfrenssen could really do with that shiny new `og_get_best_group_audience_field ` method now :)